doc/appendices/development.rst

```diff
@@ -8,9 +8,63 @@
 Development Guide
 *****************
 
+The agility, extensibility, and performance of Celeritas depend strongly on
+software infrastructure and best practices. This appendix describes how to
+modify and extend the codebase.
+
+.. include:: ../../CONTRIBUTING.rst
+
+
+.. _code_guidelines:
+
+Code development guidelines
+===========================
+
+Every new piece of code is a commitment for you and other developers to
+maintain it in the future (or delete it if obsolete). There are numerous
+considerations to making code easier to update or understand, including testing
+and documentation.
+
+
+Document implicitly and explicitly
+----------------------------------
+
+Code should be self-documenting as far as possible (see details below for
+naming conventions). This means that variable names, function names, and
+function arguments should be as "obvious" as possible. Take particular care
+with constants that appear in physics implementations. They should
+be multiplied by units in the native Celeritas unit system if applicable, or
+defined as ``Quantity`` instances. The numerical value of the constant must
+also be documented with a paper citation or other comment.
+
+
+Test thoroughly
+---------------
+
+Functions should use programmatic assertions whenever assumptions are made:
```

Comment:
You may want to explicitly state here which of the macro are active only for debug build and which are always on.